File: Cpp/394_Decode_String.cpp
class Solution {
public:
    string decodeString(string s) {
        stack<int> num_stack;
        stack<string> word_stack;
        string word = "";
        int num = 0;
        
        for(auto& ch : s){
            if(isdigit(ch)){
                num = num * 10 + ch - '0';
            }else if(ch == '['){
                word_stack.push(word);
                num_stack.push(num);
                num = 0;
                word = "";
            }else if(ch == ']'){
                int n = num_stack.top();
                string temp = word; 
                for(int i=0; i<n-1; ++i){
                    word += temp;
                }
                
                word = word_stack.top() + word;
                word_stack.pop();
                num_stack.pop();
            }else {
                word += ch;
            }
        }
        
        return word;
    }
};

File: Kotlin/394_Decode_String.kt
class Solution {
    fun decodeString(s: String): String {
        val wordStack = Stack<StringBuilder>()
        val numStack = Stack<Int>()
        val result = StringBuilder()
        var word = StringBuilder()
        var num = 0
        
        for(ch in s){
            if(ch.isDigit()){
                num = num * 10 + (ch - '0')
                continue
            }else if(ch == '['){
                wordStack.push(word)
                numStack.push(num)
                num = 0
                word = StringBuilder()
            }else if(ch == ']'){
                val n = numStack.peek()
                
                var temp = StringBuilder()
                temp.append(word)
                
                for(i in 0 until n - 1){
                    word.append(temp)
                }
                
                word = wordStack.peek().append(word)
                wordStack.pop()
                numStack.pop()
            }else {
                word.append(ch)
            }
        }
        
        return word.toString()
    }
}